chronos/streams/tlsstream.nim

```diff
@@ -0,0 +1,334 @@
+    code = int(message)
+  elif T is int:
+    msg.add(sslErrorMsg(message) & " (code: " & $message & ")")
+    code = message
+  else:
+    msg.add("Internal Error")
+  var err = newException(TlsStreamProtocolError, msg)
+  err.errCode = code
+  err
+
+# proc raiseTlsStreamProtoError*[T](message: T) =
+#   raise newTlsStreamProtocolError(message)
+
+proc getStringState*(state: cuint): string =
+  var n = newSeq[string]()
+  if (state and SSL_CLOSED) == SSL_CLOSED:
+    n.add("Closed")
+  if (state and SSL_SENDREC) == SSL_SENDREC:
+    n.add("SendRec")
+  if (state and SSL_RECVREC) == SSL_RECVREC:
+    n.add("RecvRec")
+  if (state and SSL_SENDAPP) == SSL_SENDAPP:
+    n.add("SendApp")
+  if (state and SSL_RECVAPP) == SSL_RECVAPP:
+    n.add("RecvApp")
+  result = "{" & n.join(", ") & "} number (" & $state & ")"
+
+proc tlsWriteLoop(stream: AsyncStreamWriter) {.async.} =
+  var wstream = cast[TlsStreamWriter](stream)
+  var engine: ptr SslEngineContext
+  var error: ref Exception
+
+  if wstream.kind == TLSStreamKind.Server:
+    engine = addr wstream.scontext.eng
+  else:
+    engine = addr wstream.ccontext.eng
+
+  wstream.state = AsyncStreamState.Running
+
+  try:
+    var length: uint
+    while true:
+      var state = engine.sslEngineCurrentState()
+
+      if (state and SSL_CLOSED) == SSL_CLOSED:
+        wstream.state = AsyncStreamState.Finished
+        break
+
+      if (state and (SSL_RECVREC or SSL_RECVAPP)) != 0:
+        if not(wstream.switchToReader.isSet()):
+          wstream.switchToReader.fire()
+
+      if (state and (SSL_SENDREC or SSL_SENDAPP)) == 0:
+        await wstream.switchToWriter.wait()
+        wstream.switchToWriter.clear()
+        # We need to refresh `state` because we just returned from readerLoop.
+        continue
+
+      if (state and SSL_SENDREC) == SSL_SENDREC:
+        # TLS record needs to be sent over stream.
+        length = 0'u
+        var buf = sslEngineSendrecBuf(engine, length)
+        doAssert(length != 0 and not isNil(buf))
+        var fut = awaitne wstream.wsource.write(buf, int(length))
+        if fut.failed():
+          error = fut.error
+          break
+        sslEngineSendrecAck(engine, length)
+        continue
+
+      if (state and SSL_SENDAPP) == SSL_SENDAPP:
+        # Application data can be sent over stream.
+        var item = await wstream.queue.get()
+        if item.size > 0:
+          length = 0'u
+          var buf = sslEngineSendappBuf(engine, length)
+          let toWrite = min(int(length), item.size)
+
+          if int(length) >= item.size:
+            if item.kind == Pointer:
+              let p = cast[pointer](cast[uint](item.data1) + uint(item.offset))
+              copyMem(buf, p, item.size)
+            elif item.kind == Sequence:
+              copyMem(buf, addr item.data2[item.offset], item.size)
+            elif item.kind == String:
+              copyMem(buf, addr item.data3[item.offset], item.size)
```

Comment:
These lines are encoding logic that's specific to the `item` type. It seems to me that they'll have to be duplicated if there is another place where an async stream queue item must be written to memory buffer of limited size, so they can be extracted in a helper function.

Furthermore, the limited size is captured in `toWrite`, but this variable is not used elsewhere in this function?

Reply:
@zah maybe because you started to review too early? :) Not all ideas is implemented.